feat: add `blas/ext/base/ddiff` (#10376)
stdlib-js/stdlib: lib/node_modules/@stdlib/blas/ext/base/ddiff/README.md

<!--

@license Apache-2.0

Copyright (c) 2026 The Stdlib Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

-->

# ddiff

> Calculate the k-th discrete forward difference of a double-precision floating-point strided array.

<section class="usage">

## Usage

```javascript
var ddiff = require( '@stdlib/blas/ext/base/ddiff' );
```

<!-- lint disable maximum-heading-length -->

#### ddiff( N, k, x, strideX, N1, prepend, strideP, N2, append, strideA, out, strideOut, workspace, strideW )

> Calculates the k-th discrete forward differences of a double-precision floating-point strided array.

Reviewer: ```suggestion
Calculates the k-th discrete forward difference of a double-precision floating-point strided array.
```

Be consistent. You are inconsistent throughout this PR.

```javascript
var Float64Array = require( '@stdlib/array/float64' );

var x = new Float64Array( [ 2.0, 4.0, 6.0, 8.0, 10.0 ] );
var p = new Float64Array( [ 1.0 ] );
var a = new Float64Array( [ 11.0 ] );
var out = new Float64Array( 6 );
var w = new Float64Array( 6 );

ddiff( x.length, 1, x, 1, 1, p, 1, 1, a, 1, out, 1, w, 1 );

console.log( out );
// out => <Float64Array>[ 1.0, 2.0, 2.0, 2.0, 2.0, 1.0 ]
```

The function has the following parameters:

-   **N**: number of indexed elements.
-   **k**: number of times to recursively compute differences.
-   **x**: input [`Float64Array`][@stdlib/array/float64].
-   **strideX**: stride length for `x`.
-   **N1**: number of elements to `prepend`.
-   **prepend**: a [`Float64Array`][@stdlib/array/float64] containing values to prepend prior to computing differences.
-   **strideP**: stride length for `prepend`.
-   **N2**: number of elements to `append`.

> -   **append**: a [`Float64Array`][@stdlib/array/float64] containing values to append prior to computing differences.

Reviewer: ```suggestion
-   **N1**: number of indexed elements to `prepend`.
-   **prepend**: a [`Float64Array`][@stdlib/array/float64] containing values to prepend prior to computing differences.
-   **strideP**: stride length for `prepend`.
-   **N2**: number of indexed elements to `append`.
-   **append**: a [`Float64Array`][@stdlib/array/float64] containing values to append prior to computing differences.
```

var p = new Float64Array( [ 1.0 ] );
var a = new Float64Array( [ 11.0 ] );
var out = new Float64Array( 4 );
var w = new Float64Array( 4 );

ddiff( 3, 1, x, 2, 1, p, 1, 1, a, 1, out, 1, w, 1 );

console.log( out );
// out => <Float64Array>[ 1.0, 4.0, 4.0, 1.0 ]
```

Note that indexing is relative to the first index. To introduce an offset, use [`typed array`][mdn-typed-array] views.

```javascript
var Float64Array = require( '@stdlib/array/float64' );

// Initial array...
var x0 = new Float64Array( [ 2.0, 4.0, 6.0, 8.0, 10.0 ] );

// Create an offset view...
var x1 = new Float64Array( x0.buffer, x0.BYTES_PER_ELEMENT*1 ); // start at 2nd element

var p = new Float64Array( [ 1.0 ] );
var a = new Float64Array( [ 11.0 ] );
var out = new Float64Array( 5 );
var w = new Float64Array( 5 );

ddiff( x1.length, 1, x1, 1, 1, p, 1, 1, a, 1, out, 1, w, 1 );

console.log( out );
// out => <Float64Array>[ 3.0, 2.0, 2.0, 2.0, 1.0 ]
```

<!-- lint disable maximum-heading-length -->

#### ddiff.ndarray( N, k, x, strideX, offsetX, N1, prepend, strideP, offsetP, N2, append, strideA, offsetA, out, strideOut, offsetOut, workspace, strideW, offsetW )

> Calculates the k-th discrete forward differences of a double-precision floating-point strided array using alternative indexing semantics.

Reviewer: ```suggestion
Calculates the k-th discrete forward difference of a double-precision floating-point strided array using alternative indexing semantics.
```

```javascript
var Float64Array = require( '@stdlib/array/float64' );

var x = new Float64Array( [ 2.0, 4.0, 6.0, 8.0, 10.0 ] );
var p = new Float64Array( [ 1.0 ] );
var a = new Float64Array( [ 11.0 ] );
var out = new Float64Array( 6 );
var w = new Float64Array( 6 );

ddiff.ndarray( x.length, 1, x, 1, 0, 1, p, 1, 0, 1, a, 1, 0, out, 1, 0, w, 1, 0 );

console.log( out );
// out => <Float64Array>[ 1.0, 2.0, 2.0, 2.0, 2.0, 1.0 ]
```

The function has the following additional parameters:

-   **offsetX**: starting index for `x`.
-   **offsetP**: starting index for `prepend`.
-   **offsetA**: starting index for `append`.
-   **offsetOut**: starting index for `out`.
-   **offsetW**: starting index for `workspace`.

> While [`typed array`][mdn-typed-array] views mandate a view offset based on the underlying buffer, the offset parameter supports indexing semantics based on a starting index. For example, to access only the last three elements:

Reviewer: ```suggestion
While [`typed array`][mdn-typed-array] views mandate a view offset based on the underlying buffer, the offset parameters support indexing semantics based on starting indices. For example, to access only the last three elements:
```

```javascript
var Float64Array = require( '@stdlib/array/float64' );

var x = new Float64Array( [ 2.0, 4.0, 6.0, 8.0, 10.0 ] );
var p = new Float64Array( [ 1.0 ] );
var a = new Float64Array( [ 11.0 ] );
var out = new Float64Array( 4 );
var w = new Float64Array( 4 );

ddiff.ndarray( 3, 1, x, 1, x.length-3, 1, p, 1, 0, 1, a, 1, 0, out, 1, 0, w, 1, 0 );

console.log( out );
// out => <Float64Array>[ 5.0, 2.0, 2.0, 1.0 ]
```

</section>

<!-- /.usage -->

<section class="notes">

## Notes

> -   If `N <= 0`, both functions return the output array unchanged.

Reviewer: ```suggestion
-   When `k <= 1`, the workspace array is unused and thus ignored.
-   If `N <= 0`, both functions return the output array unchanged.
```

var ddiff = require( '@stdlib/blas/ext/base/ddiff' );

var x = discreteUniform( 10, -100, 100, {
    'dtype': 'float64'
});
console.log( 'Input array: ', x );

var p = discreteUniform( 2, -100, 100, {
    'dtype': 'float64'
});
console.log( 'Prepend array: ', p );

var a = discreteUniform( 2, -100, 100, {
    'dtype': 'float64'
});
console.log( 'Append array: ', a );

var out = new Float64Array( 10 );
var w = new Float64Array( 13 );

ddiff( x.length, 4, x, 1, 2, p, 1, 2, a, 1, out, 1, w, 1 );
console.log( 'Output: ', out );
```

</section>

<!-- /.examples -->

<!-- C interface documentation. -->

* * *

<section class="c">

## C APIs

<!-- Section to include introductory text. Make sure to keep an empty line after the intro `section` element and another before the `/section` close. -->

<section class="intro">

</section>

<!-- /.intro -->

<!-- C usage documentation. -->

<section class="usage">

### Usage

```c
#include "stdlib/blas/ext/base/ddiff.h"
```

<!-- lint disable maximum-heading-length -->

#### stdlib_strided_ddiff( N, k, \*X, strideX, N1, \*prepend, strideP, N2, \*append, strideA, \*out, strideOut, \*workspace, strideW )

> Calculates the k-th discrete forward differences of a double-precision floating-point strided array.

Reviewer: ```suggestion
Calculates the k-th discrete forward difference of a double-precision floating-point strided array.
```

```c
const double x[] = { 2.0, 4.0, 6.0, 8.0, 10.0 };
const double p[] = { 1.0 };
const double a[] = { 11.0 };
double out[] = { 0.0, 0.0, 0.0, 0.0, 0.0, 0.0 };
double w[] = { 0.0, 0.0, 0.0, 0.0, 0.0, 0.0 };

stdlib_strided_ddiff( 5, 1, x, 1, 1, p, 1, 1, a, 1, out, 1, w, 1 );
```

The function accepts the following arguments:

-   **N**: `[in] CBLAS_INT` number of indexed elements.
-   **k**: `[in] CBLAS_INT` number of times to recursively compute differences.
-   **X**: `[in] double*` input array.
-   **strideX**: `[in] CBLAS_INT` stride length for `X`.
-   **N1**: `[in] CBLAS_INT` number of elements in `Prepend`.
-   **Prepend**: `[in] double*` array containing values to prepend prior to computing differences.
-   **strideP**: `[in] CBLAS_INT` stride length for `Prepend`.

> -   **N2**: `[in] CBLAS_INT` number of elements in `Append`.

Reviewer: ```suggestion
-   **N1**: `[in] CBLAS_INT` number of indexed elements for `Prepend`.
-   **Prepend**: `[in] double*` array containing values to prepend prior to computing differences.
-   **strideP**: `[in] CBLAS_INT` stride length for `Prepend`.
-   **N2**: `[in] CBLAS_INT` number of indexed elements for `Append`.
```

-   **append**: `[in] double*` array containing values to append prior to computing differences.
-   **strideA**: `[in] CBLAS_INT` stride length for `append`.
-   **out**: `[inout] double*` output array. Must have `N + N1 + N2 - k` elements.
-   **strideOut**: `[in] CBLAS_INT` stride length for `out`.
-   **workspace**: `[inout] double*` workspace array. Must have `N + N1 + N2 - 1` elements.
-   **strideW**: `[in] CBLAS_INT` stride length for `workspace`.

```c
void stdlib_strided_ddiff( const CBLAS_INT N, const CBLAS_INT k, const double *X, const CBLAS_INT strideX, const CBLAS_INT N1, const double *prepend, const CBLAS_INT strideP, const CBLAS_INT N2, const double *append, const CBLAS_INT strideA, double *out, const CBLAS_INT strideOut, double *workspace, const CBLAS_INT strideW );
```

<!-- lint disable maximum-heading-length -->

#### stdlib_strided_ddiff_ndarray( N, k, \*X, strideX, offsetX, N1, \*prepend, strideP, offsetP, N2, \*append, strideA, offsetA, \*out, strideOut, offsetOut, \*workspace, strideW, offsetW )

> Calculates the k-th discrete forward differences of a double-precision floating-point strided array using alternative indexing semantics.

Reviewer: ```suggestion
Calculates the k-th discrete forward difference of a double-precision floating-point strided array using alternative indexing semantics.
```

```c
const double x[] = { 2.0, 4.0, 6.0, 8.0, 10.0 };
const double p[] = { 1.0 };
const double a[] = { 11.0 };
double out[] = { 0.0, 0.0, 0.0, 0.0, 0.0, 0.0 };
double w[] = { 0.0, 0.0, 0.0, 0.0, 0.0, 0.0 };

stdlib_strided_ddiff_ndarray( 5, 1, x, 1, 0, 1, p, 1, 0, 1, a, 1, 0, out, 1, 0, w, 1, 0 );
```

The function accepts the following arguments:

-   **N**: `[in] CBLAS_INT` number of indexed elements.
-   **k**: `[in] CBLAS_INT` number of times to recursively compute differences.
-   **X**: `[in] double*` input array.
-   **strideX**: `[in] CBLAS_INT` stride length for `X`.
-   **offsetX**: `[in] CBLAS_INT` starting index for `X`.
-   **N1**: `[in] CBLAS_INT` number of elements in `Prepend`.
-   **Prepend**: `[in] double*` array containing values to prepend prior to computing differences.
-   **strideP**: `[in] CBLAS_INT` stride length for `Prepend`.
-   **offsetP**: `[in] CBLAS_INT` starting index for `Prepend`.

> -   **N2**: `[in] CBLAS_INT` number of elements in `Append`.

Reviewer: ```suggestion
-   **N1**: `[in] CBLAS_INT` number of indexed elements for `Prepend`.
-   **Prepend**: `[in] double*` array containing values to prepend prior to computing differences.
-   **strideP**: `[in] CBLAS_INT` stride length for `Prepend`.
-   **offsetP**: `[in] CBLAS_INT` starting index for `Prepend`.
-   **N2**: `[in] CBLAS_INT` number of indexed elements for `Append`.
```

-   **strideW**: `[in] CBLAS_INT` stride length for `Workspace`.
-   **offsetW**: `[in] CBLAS_INT` starting index for `Workspace`.

```c
void stdlib_strided_ddiff_ndarray( const CBLAS_INT N, const CBLAS_INT k, const double *X, const CBLAS_INT strideX, const CBLAS_INT offsetX, const CBLAS_INT N1, const double *Prepend, const CBLAS_INT strideP, const CBLAS_INT offsetP, const CBLAS_INT N2, const double *Append, const CBLAS_INT strideA, const CBLAS_INT offsetA, double *Out, const CBLAS_INT strideOut, const CBLAS_INT offsetOut, double *Workspace, const CBLAS_INT strideW, const CBLAS_INT offsetW );
```

</section>

<!-- /.usage -->

<!-- C API usage notes. Make sure to keep an empty line after the `section` element and another before the `/section` close. -->

<section class="notes">

</section>

<!-- /.notes -->

<!-- C API usage examples. -->

<section class="examples">

### Examples

```c
#include "stdlib/blas/ext/base/ddiff.h"
#include <stdio.h>

int main( void ) {
    // Create a strided array:
    const double x[] = { 1.0, -2.0, 3.0, -4.0, 5.0, -6.0, 7.0, -8.0 };

    // Define prepend values:
    const double p[] = { -1.0 };

>     // Define append values:

Reviewer: ```suggestion
    // Define a list of values to prepend:
    const double p[] = { -1.0 };

    // Define a list of values to append:
```

Calculates the k-th discrete forward differences of a double-precision floating-point strided array.

```javascript
var Float64Array = require( '@stdlib/array/float64' );

var x = new Float64Array( [ 2.0, 4.0, 6.0, 8.0, 10.0 ] );
var p = new Float64Array( [ 1.0 ] );
var a = new Float64Array( [ 11.0 ] );
var out = new Float64Array( 6 );
var w = new Float64Array( 6 );

ddiff( x.length, 1, x, 1, 1, p, 1, 1, a, 1, out, 1, w, 1 );

console.log( out );
// out => <Float64Array>[ 1.0, 2.0, 2.0, 2.0, 2.0, 1.0 ]
```

The function has the following parameters:

-   **N**: number of indexed elements.
-   **k**: number of times to recursively compute differences.
-   **x**: input [`Float64Array`][@stdlib/array/float64].
-   **strideX**: stride length for `x`.
-   **N1**: number of elements to `prepend`.
-   **prepend**: a [`Float64Array`][@stdlib/array/float64] containing values to prepend prior to computing differences.
-   **strideP**: stride length for `prepend`.
-   **N2**: number of elements to `append`.
-   **append**: a [`Float64Array`][@stdlib/array/float64] containing values to append prior to computing differences..
-   **strideA**: strides length for `append`.

> -   **out**: output [`Float64Array`][@stdlib/array/float64].

Reviewer: ```suggestion
-   **out**: output [`Float64Array`][@stdlib/array/float64]. Must have `N + N1 + N2 - k` elements.
```

-   **strideOut**: stride length for `out`.

> -   **workspace**: workspace [`Float64Array`][@stdlib/array/float64].

Reviewer: ```suggestion
-   **workspace**: workspace [`Float64Array`][@stdlib/array/float64]. Must have `N + N1 + N2 - 1` elements.
```

var w = new Float64Array( 6 );

ddiff( x.length, 1, x, 1, 0, 1, p, 1, 0, 1, a, 1, 0, out, 1, 0, w, 1, 0 );

console.log( out );
// out => <Float64Array>[ 1.0, 2.0, 2.0, 2.0, 2.0, 1.0 ]
```

The function has the following additional parameters:

-   **offsetX**: starting index for `x`.
-   **offsetP**: starting index for `prepend`.
-   **offsetA**: starting index for `append`.
-   **offsetOut**: starting index for `out`.
-   **offsetW**: starting index of `workspace`.

While [`typed array`][mdn-typed-array] views mandate a view offset based on the underlying buffer, the offset parameter supports indexing semantics based on a starting index. For example, to access only the last three elements:

```javascript
var Float64Array = require( '@stdlib/array/float64' );

var x = new Float64Array( [ 2.0, 4.0, 6.0, 8.0, 10.0 ] );
var p = new Float64Array( [ 1.0 ] )
var a = new Float64Array( [ 11.0 ] );
var out = new Float64Array( 4 );
var w = new Float64Array( 4 );

ddiff( 3, 1, x, 1, x.length-3, 1, p, 1, 0, 1, a, 1, 0, out, 1, 0, w, 1, 0 );

console.log( out );
// out => <Float64Array>[ 5.0, 2.0, 2.0, 1.0 ]
```

</section>

<!-- /.usage -->

<section class="notes">

## Notes

> -   If `N <= 0`, both functions return `x` unchanged.

Reviewer: You'll want to list the expectations for the expected number of elements in `out` and `workspace`. This could also be included in the parameter descriptions above.

Calculates the k-th discrete forward difference of a double-precision floating-point strided array.

```c
const double x[] = { 2.0, 4.0, 6.0, 8.0, 10.0 };
const double p[] = { 1.0 };
const double a[] = { 11.0 };
double out[] = { 0.0, 0.0, 0.0, 0.0, 0.0, 0.0 };
double w[] = { 0.0, 0.0, 0.0, 0.0, 0.0, 0.0 };

stdlib_strided_ddiff( 5, 1, x, 1, 1, p, 1, 1, a, 1, out, 1, w, 1 );
```

The function accepts the following arguments:

-   **N**: `[in] CBLAS_INT` number of indexed elements.
-   **k**: `[in] CBLAS_INT` number of times to recursively compute differences.
-   **X**: `[in] double*` input array.
-   **strideX**: `[in] CBLAS_INT` stride length for `X`.
-   **N1**: `[in] CBLAS_INT` number of indexed elements for `Prepend`.
-   **Prepend**: `[in] double*` array containing values to prepend prior to computing differences.
-   **strideP**: `[in] CBLAS_INT` stride length for `Prepend`.
-   **N2**: `[in] CBLAS_INT` number of indexed elements for `Append`.
-   **Append**: `[in] double*` array containing values to append prior to computing differences.
-   **strideA**: `[in] CBLAS_INT` stride length for `Append`.

> -   **Out**: `[inout] double*` output array. Must have `N + N1 + N2 - k` elements.

Reviewer: ```suggestion
-   **Out**: `[out] double*` output array. Must have `N + N1 + N2 - k` elements.
```

-   **strideOut**: `[in] CBLAS_INT` stride length for `Out`.

> -   **Workspace**: `[inout] double*` workspace array. Must have `N + N1 + N2 - 1` elements.

Reviewer: ```suggestion
-   **Workspace**: `[out] double*` workspace array. Must have `N + N1 + N2 - 1` elements.
```

```c
double x[] = { 2.0, 4.0, 6.0, 8.0, 10.0 };
double p[] = { 1.0 };
double a[] = { 11.0 };
double out[] = { 0.0, 0.0, 0.0, 0.0, 0.0, 0.0 };
double w[] = { 0.0, 0.0, 0.0, 0.0, 0.0, 0.0 };

stdlib_strided_ddiff_ndarray( 5, 1, x, 1, 0, 1, p, 1, 0, 1, a, 1, 0, out, 1, 0, w, 1, 0 );
```

The function accepts the following arguments:

-   **N**: `[in] CBLAS_INT` number of indexed elements.
-   **k**: `[in] CBLAS_INT` number of times to recursively compute differences.
-   **x**: `[in] double` input [`Float64Array`][@stdlib/array/float64].
-   **strideX**: `[in] CBLAS_INT` stride length for `x`.
-   **offsetX**: `[in] CBLAS_INT` starting index for `x`.
-   **N1**: `[in] CBLAS_INT` number of elements in `prepend`.
-   **prepend**: `[in] double` a [`Float64Array`][@stdlib/array/float64] containing values to prepend prior to computing differences.
-   **strideP**: `[in] CBLAS_INT` stride length for `prepend`.
-   **offsetP**: `[in] CBLAS_INT` starting index for `prepend`.
-   **N2**: `[in] CBLAS_INT` number of elements in `append`.
-   **append**: `[in] double` a [`Float64Array`][@stdlib/array/float64] containing values to append prior to computing differences..

> -   **strideA**: `[in] CBLAS_INT` strides length for `append`.

Reviewer 2: Same typos here too (ndarray C section): double period and "strides length".

```suggestion
-   **append**: `[in] double` a [`Float64Array`][@stdlib/array/float64] containing values to append prior to computing differences.
-   **strideA**: `[in] CBLAS_INT` stride length for `append`.
```

-   **offsetA**: `[in] CBLAS_INT` starting index for `Append`.
-   **Out**: `[inout] double*` output array. Must have `N + N1 + N2 - k` elements.
-   **strideOut**: `[in] CBLAS_INT` stride length for `Out`.
-   **offsetOut**: `[in] CBLAS_INT` starting index for `Out`.

> -   **Workspace**: `[inout] double*` workspace array. Must have `N + N1 + N2 - 1` elements.

Reviewer: ```suggestion
-   **Out**: `[out] double*` output array. Must have `N + N1 + N2 - k` elements.
-   **strideOut**: `[in] CBLAS_INT` stride length for `Out`.
-   **offsetOut**: `[in] CBLAS_INT` starting index for `Out`.
-   **Workspace**: `[out] double*` workspace array. Must have `N + N1 + N2 - 1` elements.
```

I am not sure why these arrays were labeled as `inout`. They are purely for writing output values.

```

</section>

<!-- /.usage -->

<!-- C API usage notes. Make sure to keep an empty line after the `section` element and another before the `/section` close. -->

<section class="notes">

</section>

<!-- /.notes -->

<!-- C API usage examples. -->

<section class="examples">

### Examples

```c
#include "stdlib/blas/ext/base/ddiff.h"
#include <stdio.h>

int main( void ) {
    // Create a strided array:
    double x[] = { 1.0, -2.0, 3.0, -4.0, 5.0, -6.0, 7.0, -8.0 };

    // Define prepend values:
    double p[] = { -1.0 };

    // Define append values:
    double a[] = { 10.0 };

    // Define output array:
    double out[] = { 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0 };

    // Define workspace:
    double w[] = { 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0 };

>     // Fill the array:

Reviewer: ```suggestion
    // Compute forward differences:
```
